internal/agent/reset.go

```diff
@@ -20,13 +22,78 @@ import (
 	"github.com/mudler/go-pluggable"
 )
 
-func Reset(reboot, unattended bool, dir ...string) error {
+func Reset(reboot, unattended, resetOem bool, dir ...string) error {
+	if internalutils.UkiBootMode() == internalutils.UkiHDD {
+		return resetUki(reboot, unattended, resetOem, dir...)
+	} else {
+		return reset(reboot, unattended, resetOem, dir...)
+	}
+}
+
+func reset(reboot, unattended, resetOem bool, dir ...string) error {
+	cfg, err := sharedReset(reboot, unattended, resetOem, dir...)
+	if err != nil {
+		return err
+	}
+	// Load the installation Config from the cloud-config data
+	resetSpec, err := config.ReadResetSpecFromConfig(cfg)
+	if err != nil {
+		return err
+	}
+
+	err = resetSpec.Sanitize()
+	if err != nil {
+		return err
+	}
+
+	resetAction := action.NewResetAction(cfg, resetSpec)
+	if err = resetAction.Run(); err != nil {
+		cfg.Logger.Errorf("failed to reset uki: %s", err)
```

Comment:
I guess this one is not uki ... which relates to the point I was going to make ... reset and reset uki are so similar that it might make sense just to distinguish between readresetuki and readreset to avoid having disparity between these two functions

Reply:
yeah the problem with that is the fucking ugliness of the type casting shit we have to do.


Firs we would need to create an interface for ResetSpec that its more generic, so a ReadReset generic function could return a generic ResetSpec, then we have to cast that crap into the proper type and plead to ged that its not broken.

I dont known, I hate this metaprogramming or what the hell the name is on golang, its just not well done and involves a lot of ugly code.

I did something similar for the InstallSpec and the extra code that I had to add just for that was not worth it in my opinion: https://github.com/kairos-io/kairos-agent/blob/main/pkg/types/v1/config.go#L45

```

func (i *InstallSpec) ShouldReboot() bool                 { return i.Reboot }
func (i *InstallSpec) ShouldShutdown() bool               { return i.PowerOff }
func (i *InstallSpec) GetTarget() string                  { return i.Target }
func (i *InstallSpec) GetPartTable() string               { return i.PartTable }
func (i *InstallSpec) GetPartitions() ElementalPartitions { return i.Partitions }
func (i *InstallSpec) GetExtraPartitions() PartitionList  { return i.ExtraPartitions }

```

Madness.

Reply:
Also this already goes down the rabbit whole into a more generic function that does the same for all specs `ReadSpecFromCloudConfig` which means they are all centralized and they all do the same thing.

Its just that on the agent the ugliness is hidden and a much clear view is shown.

Reply:
oki, you have more experience with it so not a strong push from my side :P ... then the only change needed is on the error message to not say uki